File: src/robot_life/event_engine/builder.py
from __future__ import annotations
from robot_life.common.contracts import canonical_event_detected, default_event_priority
from robot_life.common.payload_contracts import DetectionPayloadAccessor
from robot_life.common.schemas import DetectionResult, EventPriority, RawEvent, new_id, now_mono


class EventBuilder:
    """Convert detector outputs into normalized raw events.

    When ``event_priorities`` is provided, the builder automatically
    resolves the priority from the event type, eliminating the need
    for callers to hardcode it.
    """

    def __init__(
        self,
        event_priorities: dict[str, EventPriority] | None = None,
    ) -> None:
        self.event_priorities: dict[str, EventPriority] = dict(event_priorities or {})

    def build(
        self,
        detection: DetectionResult,
        priority: EventPriority | None = None,
        ttl_ms: int = 3_000,
    ) -> RawEvent:
        canonical_event_type = self._canonical_event_type(detection.event_type)
        canonical_detected_type = f"{canonical_event_type}_detected"

        # Auto-resolve priority from mapping if not explicitly passed.
        resolved_priority = priority
        if resolved_priority is None:
            resolved_priority = self.event_priorities.get(canonical_detected_type)
        if resolved_priority is None:
            resolved_priority = self.event_priorities.get(canonical_event_type)
        if resolved_priority is None:
            resolved_priority = default_event_priority(canonical_detected_type)

        accessor = DetectionPayloadAccessor.from_detection(detection)
        cooldown_target = accessor.target_id or canonical_event_type
        payload = accessor.to_dict()
        payload.setdefault("event_confidence", detection.confidence)
        payload.setdefault("raw_event_type", detection.event_type)

        return RawEvent(
            event_id=new_id(),
            trace_id=detection.trace_id,
            event_type=f"{canonical_event_type}_detected",
            priority=resolved_priority,
            timestamp_monotonic=now_mono(),
            confidence=detection.confidence,
            source=detection.detector,
            ttl_ms=ttl_ms,
            cooldown_key=f"{canonical_event_type}:{cooldown_target}",
            payload=payload,
        )

    @staticmethod
    def _canonical_event_type(event_type: str) -> str:
        """Normalize detector-level event names into canonical pipeline events."""
        normalized = canonical_event_detected(event_type)
        return normalized.removesuffix("_detected")

File: Robot_Life_CPP/src/event_engine/builder.cpp
#include "robot_life_cpp/event_engine/builder.hpp"

#include <algorithm>
#include <cctype>

#include "robot_life_cpp/common/contracts.hpp"

namespace robot_life_cpp::event_engine {

EventBuilder::EventBuilder(
    std::unordered_map<std::string, common::EventPriority> event_priorities)
    : event_priorities_(std::move(event_priorities)) {}

common::RawEvent EventBuilder::build(
    const common::DetectionResult& detection,
    std::optional<common::EventPriority> priority,
    int ttl_ms) const {
  const auto canonical_type = canonical_event_type(detection.event_type);
  const auto canonical_detected = common::contracts::canonical_event_detected(canonical_type);

  common::EventPriority resolved = priority.value_or(common::EventPriority::P2);
  if (!priority.has_value()) {
    auto it = event_priorities_.find(canonical_detected);
    if (it == event_priorities_.end()) {
      it = event_priorities_.find(canonical_type);
    }
    if (it != event_priorities_.end()) {
      resolved = it->second;
    } else {
      resolved = default_event_priority(canonical_detected);
    }
  }

  common::RawEvent event{};
  event.event_id = common::new_id();
  event.trace_id = detection.trace_id;
  event.event_type = canonical_detected;
  event.priority = resolved;
  event.timestamp_monotonic = common::now_mono();
  event.confidence = detection.confidence;
  event.source = detection.detector;
  event.ttl_ms = ttl_ms;
  const auto target_it = detection.payload.find("target_id");
  const auto cooldown_target = target_it == detection.payload.end() ? canonical_type : target_it->second;
  event.cooldown_key = canonical_type + ":" + cooldown_target;
  event.payload = detection.payload;
  event.payload.emplace("event_confidence", std::to_string(detection.confidence));
  event.payload.emplace("raw_event_type", detection.event_type);
  return event;
}

std::string EventBuilder::canonical_event_type(const std::string& event_type) {
  const auto canonical_detected = common::contracts::canonical_event_detected(event_type);
  const std::string suffix = "_detected";
  if (canonical_detected.size() > suffix.size() &&
      canonical_detected.substr(canonical_detected.size() - suffix.size()) == suffix) {
    return canonical_detected.substr(0, canonical_detected.size() - suffix.size());
  }
  return canonical_detected;
}

common::EventPriority EventBuilder::default_event_priority(const std::string& event_type) {
  return common::contracts::default_event_priority(event_type);
}

}  // namespace robot_life_cpp::event_engine
